cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,403 @@
+
+    def _apply_cnots(prevh: Tuple[int, ...], currh: Tuple[int, ...]):
+        # This function applies in sequence the CNOTs from prevh and then currh. However, given
+        # that the h are sorted in Gray ordering and that some cancel each other, we can reduce
+        # the number of gates. See [4] for more details.
+
+        cnots: List[Tuple[int, int]] = []
+
+        if ladder_target:
+            cnots.extend((prevh[i], prevh[i + 1]) for i in reversed(range(len(prevh) - 1)))
+            cnots.extend((currh[i], currh[i + 1]) for i in range(len(currh) - 1))
+        else:
+            cnots.extend((prevh[i], prevh[-1]) for i in range(len(prevh) - 1))
+            cnots.extend((currh[i], currh[-1]) for i in range(len(currh) - 1))
+
+        cnots = _simplify_cnots(cnots)
+
+        for gate in (cirq.CNOT(qubits[c], qubits[t]) for c, t in cnots):
+            yield gate
+
+    previous_h: Tuple[int, ...] = ()
+    for h in sorted_hs:
+        w = hamiltonians[h]
+
+        yield _apply_cnots(previous_h, h)
+
+        if len(h) >= 1:
+            yield cirq.Rz(rads=(theta * w)).on(qubits[h[-1]])
+
+        previous_h = h
+
+    # Flush the last CNOTs.
+    yield _apply_cnots(previous_h, ())
+
+
+@value.value_equality
+class BooleanHamiltonian(raw_types.Operation):
+    """An operation that represents a Hamiltonian from a set of Boolean functions."""
+
+    def __init__(
+        self,
+        qubit_map: Dict[str, 'cirq.Qid'],
+        boolean_strs: Sequence[str],
+        theta: float,
+        ladder_target: bool,
+    ):
+        """
+        Builds an BooleanHamiltonian.
+
+        For each element of a sequence of Boolean expressions, the code first transforms it into a
+        polynomial of Pauli Zs that represent that particular expression. Then, we sum all the
+        polynomials, thus making a function that goes from a series to Boolean inputs to an integer
+        that is the number of Boolean expressions that are true.
+
+        For example, if we were using this gate for the max-cut problem that is typically used to
+        demonstrate the QAOA algorithm, there would be one Boolean expression per edge. Each
+        Boolean expression would be true iff the vertices on that are in different cuts (i.e. it's)
+        an XOR.
+
+        Then, we compute exp(j * theta * polynomial), which is unitary because the polynomial is
+        Hermitian.
+
+        Args:
+            boolean_strs: The list of Sympy-parsable Boolean expressions.
+            qubit_map: map of string (boolean variable name) to qubit.
+            theta: The list of thetas to scale the Hamiltonian.
+            ladder_target: Whether to use convention of figure 7a or 7b.
+        """
+        self._qubit_map: Dict[str, 'cirq.Qid'] = qubit_map
+        self._boolean_strs: Sequence[str] = boolean_strs
+        self._theta: float = theta
+        self._ladder_target: bool = ladder_target
```

Comment:
Why is this important for the user to know?

Reply:
I think the authors said they would provide a bound for the number of gates in a follow-up paper, but I couldn't find one. I tried to compute it myself a bit (not for too long, though) but it doesn't seem obvious. I decided to add the option as an input instead, so that it can be decided experimentally.

I expanded the comment to reflect why the option is given.

Reply:
Why wouldn't we try both and go with the smaller one?

Reply:
That comment 

> I think the authors said they would provide a bound for the number of gates in a follow-up paper, but I couldn't find one. 

From what I see in the paper that they had a remark on the proof that the commutation operations to simplify CNOTs lead to O(N') gates where N' < N Walsh operators:

<img src="https://user-images.githubusercontent.com/635073/124212026-2981a080-dabc-11eb-92fe-da8573cc9860.png" width="300px">
 

They had a completely different remark regarding the "ladder target" as you call it: 

It is figure 3 and not 7 (as you mentioned in the comment): 

<img src="https://user-images.githubusercontent.com/635073/124211939-ffc87980-dabb-11eb-8536-dca9a8299553.png" width="300px">

This tells me that we should not implement - only the 3(b) that they recommend...